### PR TITLE
 RONDB-244: Auto reconnect after temporary network failures 

### PR DESCRIPTION
**Pull request by smkniazi**

### CODE REVIEW

File: storage/ndb/rest-server/data-access-rondb/src/error-strings.h [hunk 1]
```diff
@@ -56,6 +56,11 @@ extern "C" {
 #define ERROR_030 "Failed to set lock mode."
 #define ERROR_031 "Failed to set filter."
 #define ERROR_032 "Failed to load index."
+#define ERROR_033 "RonDB connection is not open."
+#define ERROR_034 "RonDB reconnection is already inprogress."
+#define ERROR_035 "RonNDB connection and object pool is not initialized."
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/error-strings.h]: RoNDB => RonDB

File: storage/ndb/rest-server/data-access-rondb/src/error-strings.h [hunk 2]
```diff
@@ -56,6 +56,11 @@ extern "C" {
...
+#define ERROR_036 "RonNDB reconnection already in progress."
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/error-strings.h]: Same

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        LOG_DEBUG("GetNdbObject triggered reconnection");
+        Reconnect();
+      }
+
+      LOG_ERROR(ERROR_033 + std::string(" Connection State: ") + std::to_string(connection_state) +
+                std::string(" Reconnection State: ") + std::to_string(reconnection_in_progress));
+      return RS_SERVER_ERROR(ERROR_033);
+    }
+  }
+
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:174]: What is the purpose of releasing the mutex on connectionInfoMutex to immediately seize it again?

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    return RS_SERVER_ERROR(ERROR_033);
+  }
+
+  *rdrsRonDBConnection = __instance;
+  return RS_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::GetNdbObject(Ndb **ndb_object) {
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:143]: Looks like a potential Concurrency bottleneck. Probably could do with some split of mutexes.

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+    // for loss of connectivity to the cluster
+    if (status->classification == NdbError::UnknownResultError) {
+      LOG_ERROR("Detected connection loss. Triggering reconnection.");
+      Reconnect();
+    }
+  }
+}
+
+//--------------------------------------------------------------------------------------------------
+
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:220]: Here you return a RonDB_Stats object, is it a copy or by reference, if by reference you will get problems I gather

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+RS_Status RDRSRonDBConnection::Shutdown(bool end) {
+
+  LOG_DEBUG("Shutting down RonDB connection and NDB object pool");
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    stats.connection_state = DISCONNECTED;
+  }
+
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:243]: Again release of mutex immediately followed by new seize of mutex.

Comment (mronstro) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:243]: Or actually not

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.hpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2022 Hopsworks AB
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.hpp:2]: 2023

Comment (mronstro) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.hpp:2]: Actually 2023, 2023 (first copyright year and year of update)

File: storage/ndb/rest-server/data-access-rondb/src/retry_handler.cpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (C) 2022 Hopsworks AB
```
Comment (mronstro) [inline storage/ndb/rest-server/data-access-rondb/src/retry_handler.cpp]: 2023, 2023

File: storage/ndb/rest-server/rest-api-server/internal/log/logc.go [hunk 9]
```diff
@@ -40,7 +42,7 @@ func RegisterLogCallBack() {
 //export goLog
 func goLog(logMsg C.RS_LOG_MSG) {
 	level := log.Level(logMsg.level)
-	msg := C.GoString(&logMsg.message[0])
+	msg := fmt.Sprintf("RDRS native layer: %s", C.GoString(&logMsg.message[0]))
```
Comment (mronstro) [inline storage/ndb/rest-server/rest-api-server/internal/log/logc.go]: Maybe optimisation Vincent did is applicable here as well?

**Review by mronstro — DISMISSED** [review]
Main issue is that there could be some concurrency issues with mutex when getting Ndb objects

File: storage/ndb/rest-server/data-access-rondb/src/error-strings.h (hunk 1 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/error-strings.h]: fixed

File: storage/ndb/rest-server/data-access-rondb/src/error-strings.h (hunk 2 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/error-strings.h]: fixed

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp (hunk 3 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:174]: the first time it had only one lock and the next time it takes two locks

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp (hunk 4 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:143]: yes this is, however, vincent managed to perform 180k ops/sec using grpc and it is good for now. We can split the mutex in another PR

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp (hunk 5 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:220]: it  returned by copy

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp (hunk 6 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp:243]: refactored

File: storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.hpp (hunk 7 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.hpp:2]: fixed

File: storage/ndb/rest-server/data-access-rondb/src/retry_handler.cpp (hunk 8 above)
Comment (smkniazi) [reply storage/ndb/rest-server/data-access-rondb/src/retry_handler.cpp]: fixed

File: storage/ndb/rest-server/rest-api-server/internal/log/logc.go (hunk 9 above)
Comment (smkniazi) [reply storage/ndb/rest-server/rest-api-server/internal/log/logc.go]: yes, i have fixed it

**Review by smkniazi — COMMENTED** [review]
fixed according to review

**Review by mronstro — APPROVED** [review]

**Pull request merged**